New feature: send/recv message implementation added to network stack

mat-kalinowski opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
This PR is continuation of another stale PR: https://github.com/ARMmbed/mbed-os/pull/14847. I am opening this one, as I didn't have access to push to the fork created by other developer. 

I have added few changes related to unresolved comments:

- function names were changed from socket_sendmsg/socket_recvmsg to socket_sendto_control/socket_recvfrom_control.
- default implementation of this functions was provided in the NetworkStack class to not break the existing API. This was the issue that caused tests to fail in the previous PR.
- MsgHeaderIterator accesses elements on the aligned addresses

.


#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
Stale PR with this feature: 
https://github.com/ARMmbed/mbed-os/pull/14847
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@pan- 
----------------------------------------------------------------------------------------------------------------

tymoteuszblochmobica committed:
New feature: send/recv message  implementation added to network stack

mat-kalinowski committed:
Code improvements for the unresolved comments

This patch contains improvements mentioned in the unresolved PR
comments:
- function names were changed from socket_sendmsg/socket_recvmsg to
  socket_sendto_control/socket_recvfrom_control.
- default implementation of this functions was provided in the
  NetworkStack class.
- MsgHeaderIterator accesses elements on the aligned addresses.

pan- commented on connectivity/lwipstack/include/lwipstack/lwipopts.h:
```diff
@@ -311,6 +311,9 @@
 
 #endif
 
+// FIXME: Add compile time configuration and define conditionaly
```
This comment can be removed.

mat-kalinowski replied:
Done

pan- commented on connectivity/netsocket/include/netsocket/MsgHeader.h:
```diff
@@ -0,0 +1,99 @@
+#include "netsocket/nsapi_types.h"
+
+/**
+ * Allows iteration through the list of message headers received in the control parameter of the
```
We need doxygen that demonstrate how this can be used as well as individual doxygen comment for public functions.

mat-kalinowski replied:
Where would you like me to add doxygen with usage example, somewhere in the new file ?

pan- replied:
No you can add it to the class documentation. Here's and example for the Span class: https://github.com/ARMmbed/mbed-os/blob/c14396a5b619b185d344ad6dfc25f04cb330f1fc/platform/include/platform/Span.h#L74-L213

The code goes between the `@code` and `@endcode` blocks.

mat-kalinowski replied:
Done, thanks

pan- commented on connectivity/netsocket/include/netsocket/nsapi_types.h:
```diff
@@ -406,6 +408,23 @@ typedef struct nsapi_stagger_req {
 
+/** nsapi_msghdr
+ */
+typedef struct MBED_ALIGN(double) nsapi_msghdr {
```
@kjbracey-arm What is your preference for this alignment ?

pan- replied:
Actually, C++11 provide a `max_align_t` type (https://en.cppreference.com/w/c/types/max_align_t and https://en.cppreference.com/w/cpp/types/max_align_t). Let's use this instead.

mat-kalinowski replied:
Done

pan- reviewed (commented):
Thanks for adding the test and applying the change requested in the previous PR. I left comments for few improvements that should be made to that PR.

mat-kalinowski committed:
Adding documentation and minor code changes

Doxygen documentation for MsgHeaderIterator was added. Type used
for alignment of the nsapi_msghdr_t struct was changed.

mat-kalinowski committed:
Fixing documentation errors

Minor errors in the documentation were fixed. These caused failure
of the doxygen checks in the PR.

0xc0170 commented on connectivity/netsocket/tests/UNITTESTS/netsocket/NetworkStack/test_MsgHeaderIterator.cpp:
```diff
@@ -0,0 +1,105 @@
+    p_buff.pkt_info_1.hdr.len = 0;
+
+    EXPECT_TRUE(it.next() == nullptr);
+}
```
missing empty line at the end

mat-kalinowski replied:
Done

0xc0170 commented on connectivity/netsocket/source/TCPSocket.cpp:
```diff
@@ -173,6 +173,16 @@ nsapi_size_or_error_t TCPSocket::sendto(const SocketAddress &address, const void
+nsapi_size_or_error_t TCPSocket::sendto_control(const SocketAddress &address, const void *data, nsapi_size_t size,
+                                                nsapi_msghdr_t *control, nsapi_size_t control_size)
+{
+    // FIXME: Implement
```
Can you rather create an tracking issue for this and remove the comment (reference the issue in the commit message, it can be easily found then)

mat-kalinowski replied:
Done, new issue - https://github.com/ARMmbed/mbed-os/issues/15057

mat-kalinowski replied:
Issue number should be added in the amended commit message when merging this PR into master

0xc0170 reviewed (dismissed):
As this is a feature pull request, should there be referenced docs update?

mat-kalinowski committed:
Removing FIXME comment and adding empty line in the test file

mat-kalinowski committed:
Moving alignment information in the nsapi_msghdr_t

Macro MBED_ALIGN expands in C to _Alignas which can't be used in the
type declaration. This patch moves it to the first type definition
which makes this code compile properly in CPP and C.
